### ISSUE TITLE
[Feature Request] support `toJSONSchema`

### ISSUE THREAD

**Issue by kwaa** [standard-schema/standard-schema#21]
I'm currently writing [xsAI - a minimal TypeScript AI SDK](https://github.com/moeru-ai/xsai), and I can't migrate from [`@typeschema/main` ](https://github.com/decs/typeschema) to `@standard-schema/spec` due to the lack of `toJSONSchema` feature.

Maybe there could be a `~standard.toJSONSchema` method?

---

As an unofficial solution, you can now use `toJsonSchema` from [`xsschema`](https://xsai.js.org/docs/packages-top/xsschema):

```ts
import { type } from 'arktype'
import * as v from 'valibot'
import { toJsonSchema } from 'xsschema'
import { z } from 'zod'
 
const arktypeSchema = type({
  myString: 'string',
  myUnion: 'number | boolean',
}).describe('My neat object schema')

const valibotSchema = v.pipe(
  v.object({
    myString: v.string(),
    myUnion: v.union([v.number(), v.boolean()]),
  }),
  v.description('My neat object schema'),
)

const zodSchema = z.object({
  myString: z.string(),
  myUnion: z.union([z.number(), z.boolean()]),
}).describe('My neat object schema')

const arktypeJsonSchema = await toJsonSchema(arktypeSchema)
const valibotJsonSchema = await toJsonSchema(valibotSchema)
const zodJsonSchema = await toJsonSchema(zodSchema)
```

https://xsai.js.org/docs/packages-top/xsschema
https://github.com/standard-schema/standard-schema/issues/21#issuecomment-3069990920

**Comment by kwaa** [comment]
Just realized it's mentioned here too: https://github.com/standard-schema/standard-schema/issues/1#issuecomment-2065428229

**Comment by fabian-hiller** [comment]
My personal opinion is that this is out of scope at the moment, as it would force every library to implement such a function, and this is probably not possible for Valibot in terms of bundle size. But there are ideas for a common schema AST that would allow it to be converted to JSON schema. Here is more context:
- https://github.com/standard-schema/standard-schema/issues/7
- https://github.com/standard-schema/standard-schema/issues/11

As a workaround you could check the `.vendor` property of Standard Schema to apply a vendor-specific `toJsonSchema` function.

**Comment by kwaa** [comment]
> As a workaround you could check the `.vendor` property of Standard Schema to apply a vendor-specific `toJsonSchema` function.

Could this be offered as a package under the `@standard-schema` scope? If so, I'd be interested in contributing.

**Comment by notramo** [comment]
The problem with a vendor-specific JSON schema conversion option is that then the consumer libraries (e.g. form UI, HTTP APIs, etc.) would have to add dependencies on the schema libraries.

I recommend adding subtypes with additional capabilities like `StandardSchemaV1.JSONConvertable`, which would be the same as `StandardSchemaV1` but with an additional `['~standard'].toJSONSchema()` method. Then the schema libraries could provide their own extensions to it. Monolithic designs like Yup and Zod can include it by default, but modular libraries like Valibot and Superstruct could provide an `toJSONSchemaConvertable(mySchema)` method, which can be tree-shaken and separately imported. If a form validation library requires a JSON compatible schema, then one can call it like `createForm(toJSONConvertable(myValibotSchema))` with Valibot and `createForm(mySchema)` for other schema libraries that implement it by default. It would also make possible that a 3rd party library would supply the. JSON schema conversion method, e.g. `@valibot/to-json-schema`.

This design would also allow to add other subtypes, that can be combined with `&` operator. If a library needs a JSON Schema convertable schema and one that has a `[~standard].getDefaults()` method, it could specify this:
```typescript
const createForm = (schema: StandardSchemaV1.JSONConvertable & StandardSchemaV1.HasDefaults) …
```

**Comment by fabian-hiller** [comment]
Thank you for your feedback! Great idea! For now, we will probably focus on getting the current spec right first, and then we can focus on other use cases. Curious to hear what @colinhacks and @ssalbdivad think about such a method.

**Comment by ssalbdivad** [comment]
Would definitely support adding methods like this as long as they're optional.

**Comment by fabian-hiller** [comment]
We could either make it optional, or provide an additional interface so that third-party libraries can choose which features they need when passing a schema as an argument. I might be able to inject the `toJsonSchema` functionality using an import with a side effect similar to how our i18n feature works.

```ts
import * as v from 'valibot';
import '@valibot/to-json-schema/inject'; // Just an initial idea

// Now this works (without the import it throws an error)
const jsonSchema = v.string()["~standard"].toJsonSchema();
```

**Comment by ssalbdivad** [comment]
@fabian-hiller Love that 🔥 I have a type-level config set up that users can extend but I had never considered endpoints like that functioning as presets you could extend.

Still, for the spec, the property would need to be optional since we want libraries that don't support JSON schema conversion to still be compatible with the core validation interface.

Unless somehow the goals of the spec fundamentally change, I would not expect to ever add additional required properties. Libraries can just create their own extension of the base `StandardSchema` type that requires whatever optional properties they plan to use.

Maybe worth adding a utility like `StandardSchemaWithRequiredKeys<k extends OptionalSchemaKey>` for that?

**Comment by fabian-hiller** [comment]
I agree that we should be very careful about adding new properties and functionality. Instead of adding new properties for additional functionality, we could brand the current interface as the base interface and provide additional interfaces in the long run. For example, we could provide a `StandardSchemaToJsonSchema` interface that includes the `.toJsonSchema` property. That way, third-party libraries could choose what functionality they need/provide.

But for now, I strongly recommend getting the current interface right and helping other libraries integrate before adding new things to it or providing additional interfaces. I see most of the issues here on GitHub as cool ideas for the future. I am afraid Standard Schema will fail if we ignore this.

**Comment by fabian-hiller** [comment]
To everyone reading this: If you are a third-party library author who needs `toJsonSchema` functionality, please add a 👍 to this comment and share your use case.

**Comment by logaretm** [comment]
Adding my perspective here as a maintainer of https://github.com/formwerkjs/formwerk that bets on standard schemas.

I have a few use-cases as a form library that can be useful that would only be possible with a JSON schema or an AST.

### Path's `required` and `nullable` state

This is one important aspect we have, being able to tell that the path must exist and if it is nullable directly affects a few areas, I can think of two at the moment:

- **a11y:** Non-optional fields must be announced as such for screen readers, either via `required` attribute for HTML input elements or `aria-required` for other types of elements. 
- **styling:** Uniform styling based on the `required` attribute offers much better DX than checking with JS. This is a "nice to have" and we can live without it.

There might be different definitions of what "optional" is, and how it compares to `nullable` but I think exposing either `required` or `optional` and a `nullable` metadata would be sufficient to allow 3rd party libraries to make their own definition without forcing the standard schema project to define it.

### Default value extraction 

We discussed this in #11 but I no longer think this is something the core Formwerk would be doing as it has a lot of unknowns especially when dealing with discriminated unions. Also the confusion around Input values not exactly matching the output values would be a concern.

However, this can be important in form generators, more on that below.

### Form Generator

This is within our scope of work and we probably won't start working on it for a while unless an AST or `toJSON` exists. Similar to [autoform](https://www.shadcn-vue.com/docs/components/auto-form) that works with Zod at the moment.

Form generators require a JSON representation because they might be stored on the client or on the server. And being able to generate a form out of a Zod/Valibot/Arktype schema would be very cool, and would allow for dream-like DX workflows in that area.

There is still the tricky part about whether a path is represented by a field or not. I think this can be offloaded to the 3rd party library to make its own way of doing that to keep this spec lean. For example, an object with these two paths: `users.0.name` and `users.0.email` can be represented by either a single input component that emits an object like a select element. Or two text inputs for each leaf.

I'm thinking as 3rd party library authors, we can introduce a "mark" API that defines a path as a field or not. Or let users decide on how to build the heuristics for it with a custom predicate function or whatever.

Sorry for yapping, just a little excited for this.

--------------

I think https://github.com/standard-schema/standard-schema/issues/21#issuecomment-2557384786 is a good compromise to keep bundle sizes small and can be easily recommended by 3rd library authors in their docs when the AST is required.

**Comment by fabian-hiller** [comment]
Thank you for your feedback! I will make sure to discuss the JSON Schema functionality in more detail before releasing the final Standard Schema v1 specification. If [my idea](https://github.com/standard-schema/standard-schema/issues/21#issuecomment-2557384786) works, and most schema libraries interested in Standard Schema can and want to support this functionality, we may add it as a required property, an optional property, or as an additional interface. If we decide to include it somehow in the spec, it will probably take some time to get the details right, as I think some configuration will be needed to decide how `.toJsonSchema` should behave if a schema property cannot be converted to JSON Schema.

**Comment by fabian-hiller** [comment]
I will share the result of our discussion about JSON Schema support soon.

**Comment by fabian-hiller** [comment]
Sorry for the delay. Here are our thoughts. In general, we all see value in supporting something like a JSON schema conversion function. For example, for AI SDKs to generate structured output, and for form libraries to initialize form state or perform other similar operations. But we also noticed that JSON schema is not needed for validation purposes when using standard schema. Instead, it is used more or less as a common AST to represent a schema. The problem with JSON Schema is that it can only represent a small part of what ArkType, Valibot, Zod, and others offer. Therefore, a JSON Schema conversion function will often return incorrect results or throw an error, depending on how we would implement it. And this is where it gets very tricky... how should it be implemented? Should it be somehow configurable because different use cases require different results? What should such a function output if a schema uses a transformation or a validation JSON Schema does not support? Is JSON Schema even a good idea for a AST-like structure or would it be better to define a new AST optimized for this use case? Because of all these questions, we decided not to implement it... but we encourage library authors, especially of popular libraries, to reach out and talk to us to figure out all these questions in the long run.

**Comment by florian-lefebvre** [comment]
Hi! I'm considering adding support for Support Schema in Astro for user facing APIs, and we would need json schema generation for content layer related things:
- If we can't infer a schema from a loader, we generate a json schema to approximate it
- We need a json schema for intellisense too (`$schema` property)

**Comment by fabian-hiller** [comment]
Hi Florian, thanks for your feedback! Do you use JSON Schema for validation or just as an AST to represent a specific data structure or schema to build features like IntelliSense on top of? If so, do you encounter any limitations with JSON Schema in this regard? What do you currently do when a Zod schema is provided that cannot be fully represented with JSON Schema?

**Comment by florian-lefebvre** [comment]
We do not use the json schema for validation, purely for intellisense. I realized I said something wrong, we use `zod-to-ts` in the content layer for some specific cases where we try to "guess" the type ([source](https://github.com/withastro/astro/blob/736b9b9d721082eb231f8da64259c399d7fea1f6/packages/astro/src/content/types-generator.ts#L401-L426)).

As for limitations, using `zod-to-ts` may result in `any`s when it can't compute the type, which is not ideal.

**Comment by productdevbook** [comment]
We scan the schema with CLI, and if the schema is Zod or similar, we try to generate an object accordingly. If this feature is implemented, it seems like Standard Schema will fully achieve its purpose.

<details><summary>Example cli validation object plugins</summary>
<p>

```
const plugins: Record<ValidationLibrary, UriSchemaValidatorPlugin> = {
  zod: {
    name: 'zod',
    isSchemaType: (obj: any) => obj?._def?.typeName !== undefined,
    getRouterParams: (routerObj: any) => {
      try {
        const shape = routerObj?.shape?.params?.shape
        return shape ? Object.keys(shape) : null
      }
      catch {
        return null
      }
    },
    internalKeys: ['_def', 'shape'], // Zod-specific internal keys
  },
...
```
</p>
</details>

**Comment by fabian-hiller** [comment]
@florian-lefebvre Why is `zod-to-ts` needed? Can't you just derive the types directly from Zod, or do you need a stringified version to make it work?

**Comment by fabian-hiller** [comment]
@productdevbook How do you feel about [this comment](https://github.com/standard-schema/standard-schema/issues/21#issuecomment-2611231499)? Do you have any specific feedback that could help us answer the open questions in the long run?

**Comment by florian-lefebvre** [comment]
> @florian-lefebvre Why is `zod-to-ts` needed? Can't you just derive the types directly from Zod, or do you need a stringified version to make it work?

@ascorbic could you explain why we need this?

**Comment by ascorbic** [comment]
@fabian-hiller @florian-lefebvre Yes, we need a stringified one because the schema is returned by content loaders at build time, so that it can be constructed on the fly. This lets them do introspection on an API, for example. This means it can't use inferred types.

**Comment by fabian-hiller** [comment]
@ascorbic Have you investigated whether it is possible to infer the type produced by Zod at build time and use TS directly to convert it to a string? That way you could remove the dependency on a schema library specific conversion. I think TypeScript has a transformation feature that this functionality could be built on top of.

**Comment by florian-lefebvre** [comment]
We have been discussing this a little bit and we may change how we handle that part to not rely on `zod-to-ts`. We won't look deep into it just yet, but I plan on working on it in Q2 so I'll probably come back with more feedback/questions

**Comment by kwaa** [comment]
> > As a workaround you could check the `.vendor` property of Standard Schema to apply a vendor-specific `toJsonSchema` function.
> 
> Could this be offered as a package under the `@standard-schema` scope? If so, I'd be interested in contributing.

I released a library that supports `toJSONSchema()` - [xsschema](https://npmjs.com/package/xsschema), which calls different packages by detecting the vendor, currently only `zod` and `valibot` are supported, and it provides an interface similar to [typeschema](https://typeschema.com).

[(source)](https://github.com/moeru-ai/xsai/tree/main/packages/xsschema)

**Comment by fabian-hiller** [comment]
Great idea! Cool to see someone working on an AI SDK with a smaller footprint. Feel free to report back if you encounter any problems or challenges.

**Comment by remorses** [comment]
I would love this feature too, my use case is automatic openapi generation for [Spiceflow](https://github.com/remorses/spiceflow), a type safe API framework

**Comment by Pukimaa** [comment]
While, I'm not a maintainer of any libaries that could make use of this, I do use a library that I want to add support for it - Elysia.js. From what I know about the codebase, they mostly use `toJSONSchema()`, to allow types being passed to plugins for stuff like an openapi & swagger plugin.

**Comment by ericallam** [comment]
We have a few features at trigger.dev that require JSON schema support, and it would greatly increase our chance of adopting Standard Schema if we at least knew JSON schema support was coming: 

- Users define tasks and are able to specify a schema on that task that defines the task's "payload" type. We index tasks and provide a UI for testing tasks. We'd love to auto-generate example payloads, based on a JSON schema derived from the payload schema. We could also use the JSON schema for server side validation of payloads, to prevent invalid payloads from executing tasks. Currently we only validate on the "client" side, causing unnecessary work (we'd like to do both)
- We're working on a new feature that will enable "human in the loop" experiences where users can pause a task and wait for a token to be resolved, and we'd like to be able to pass a schema to validate the token payload. We need a JSON schema to be able to do similar things as the tasks point above: validate token payloads on the server, provide a nice UI for completing a token with a payload (e.g. a form)

**Comment by qcho** [comment]
@harrysolovay is creating a `standard-json-schema` lib that looks very promising

https://github.com/harrysolovay/standard-json-schema

**Comment by kwaa** [comment]
> **[`standard-json-schema`](https://github.com/harrysolovay/standard-json-schema) is 635 bytes minified / with types stripped out.**

It doesn't make much sense to compare the size of the packages themselves, since `effect`, `zod`, `valibot` will all be installed (even if you only need to use one of them in most cases)

> Regarding peer vs. optional deps, I believe peer is better, as it spares the developer of needing to explicitly install the JSON-schema-producing library to which `standard-json-schema` delegates. I doubt any developer will notice a difference in installation times between peer/optional, even before the initial installation/caching.

I didn't mention the `optionalDependencies` field, but rather the [`optional` in `peerDependenciesMeta`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta).

**Comment by productdevbook** [comment]
> It doesn't make much sense to compare the size of the packages themselves, since `effect`, `zod`, `valibot` will all be installed (even if you only need to use one of them in most cases)
> 

I agree with this, the system should be toJSON before packages are installed.

**Comment by harrysolovay** [comment]
> I didn't mention the optionalDependencies field, but rather the [optional in peerDependenciesMeta](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta).

@kwaa interesting! I didn't know this existed––thank you for pointing it out. Just incorporated your feedback into [`standard-json-schema` v0.1.1](https://www.npmjs.com/package/standard-json-schema/v/0.1.1).

> I agree with this, the system should be toJSON before packages are installed.

@productdevbook, can you please clarify what you mean by this? Do you mean calling `toJSON`/`JSON.stringify` on the runtime type itself? (for most libs, this would not produce a JSON schema).

**Comment by fabian-hiller** [comment]
Thank you all for helping making Standard Schema better! @MathurAditya724 also dropped a [`standard-json`](https://github.com/standard-community/standard-json) package inspired by @kwaa. Once we know what's the best implementation we could also consider releasing it as an official package.

**Comment by MathurAditya724** [comment]
> https://pkg-size.dev/standard-json-schema - 30MB install size, 1.1MB minified
> https://pkg-size.dev/xsschema - 13KB install size, 1.1KB minified

Just adding this here - 

https://pkg-size.dev/@standard-community/standard-json - 22KB install size, 1.4KB minified (supports CJS and ESM)

If we drop the support for CJS, we will be at ~ 11KB install size

**Comment by productdevbook** [comment]
Our expectation is not to install a 3rd party npm package. It was that each repository would handle it internally, like here. Why is everyone trying to make a separate npm package ? That's what I've been waiting for. I don't like the current solutions. `~standard.toJSONSchema()`

**Comment by MathurAditya724** [comment]
Simply put not all libs that supports Standard Schema, have a `toJSONSchema` function. While `arktype` has a native support for `toJSON`, `zod` relies on 3rd party lib and `valibot` has a separate package for this (official support). But even with that, the generate schema is not same across these 3 libs (they are slightly different). As for other libs, they might not think or want to put the time and effort on this.

IMHO, I am okay with this as Standard Schema project is more about a standardised validation solution, which is the main goal for all the validation libs out there.

This is something only a few people like lib/framework authors might want, and having to install a few dependencies for it, seems okay. If you don't want your users needing to install it separately, you can just add it into your peer dependency and don't mark them optional.

**Comment by fabian-hiller** [comment]
I agree that something like `.toJsonSchema` property would be best but I also agree that it is very hard to get right as described [here](https://github.com/standard-schema/standard-schema/issues/21#issuecomment-2611231499). 😐

**Comment by harrysolovay** [comment]
In case a multi-JSON-schema-producer wrapper lib is out of scope for the Standard Schema org, I just want to clarify that I’m committed to maintaining and evolving [`standard-json-schema`](https://github.com/harrysolovay/standard-json-schema). I'll actively expand support for additional runtime type libraries as they emerge and/or add JSON schema generation features. 

@kwaa while I appreciate the comparison, I think the ecosystem benefits from having multiple robust options. We can knowledge-share and push each other to improve our libraries. Locking into a single implementation might stifle that.

**Comment by mmkal** [comment]
Another use case for toJsonSchema: [trpc-cli](https://github.com/mmkal/trpc-cli). Right now it just has hand-written implementations for zod, arktype, valibot and effect.

Curious to find out which of xsschema, standard-json-schema, and typeschema is considered the "best" for which use cases and/or should be considered most reliable. typeschema seems to support a lot more libraries?

**Comment by transitive-bullshit** [comment]
I'd love to see a standardized solution using https://github.com/quansync-dev/quansync to make `toJsonSchema` sync or async.

**Comment by florian-lefebvre** [comment]
Hi again! We discussed adopting Standard Schema in Astro 6 (https://github.com/withastro/roadmap/discussions/1174) but the conclusion is we'd need json schema to be part of the spec before we can do so

**Comment by fabian-hiller** [comment]
@transitive-bullshit it what case would `toJsonSchema` be async? Is there any schema library out there where the JSON Schema generation is async?

**Comment by fabian-hiller** [comment]
@colinhacks and @ssalbdivad what is your current take on supporting an optional `.toJSONSchema` property?

**Comment by kwaa** [comment]
> [@transitive-bullshit](https://github.com/transitive-bullshit) it what case would `toJsonSchema` be async? Is there any schema library out there where the JSON Schema generation is async?

The xsschema-like solution needs to import the package corresponding to the vendor dynamically and is therefore asynchronous.

**Comment by harrysolovay** [comment]
I think this task first requires defining the minimal common subset of JSON schema that needs to be supported. Ie. desired kinds of unions (`anyOf`, `oneOf`, `enum`, ...), whether additional properties are disallowed, whether to support circularity / defs/refs. The list goes on.

I'd strongly recommend that this JSON schema subset includes only what is most widely supported by LLM providers, as this is panning out to be one of the main use cases for JSON schema.

**Comment by fabian-hiller** [comment]
@harrysolovay as a TypeScript interface specification, we have no direct control over the JSON schema output. We could limit the return type, but many schema libraries would likely not support it. I think it is unlikely that schema libraries will change their JSON Schema output just to match the spec if we define specific rules. Therefore, it seems more practical to type it as `toJSONSchema?: () => Record<string, unknown>` and ask consumers of the JSON Schema output to validate it if they only support a subset of JSON Schema. I am happy to receive feedback on my thoughts.

**Comment by finom** [comment]
Some kind of an interface that offers standard way to get JSON Schema would allow to create universal libraries for LLM agents `tool(lib.struct({ ... }))`. The `tool` function can extract parameters schema using standard interface, supporting any library that implements Standard Schema and the `toJSONSchema` function.

**Comment by finom** [comment]
What do you think of this?

```ts
schema['~standard'].toJSONSchema(input, { target: string });
```

I think `target` option would be an important feature that allows to define version of output JSON Schema. IMO it should be also present at the standard. WDYT?

**Comment by fabian-hiller** [comment]
Yes, we will probably standardize the JSON Schema options over time, and `target` could be a good place to start. What behavior would you expect if a user selects a target that the underlying schema library does not support?

To put this in context, I would prioritize adding `toJSONSchema` to the spec much more than standardizing the options. This could also be expanded at a later date.

**Comment by remorses** [comment]
This feature would also be helpful to implement [type-safe search params in `react-router`](https://github.com/remix-run/react-router/discussions/13800)

**Comment by rattrayalex** [comment]
Heya @fabian-hiller! 😄  

We're excited to support standard-schema natively in the OpenAI & Anthropic TS packages, rather than just zod, once this is ready!

**Comment by FlatMapIO** [comment]
I think this is the top priority at the moment, as it is blocking the progress of the AI ecosystem adopting the standard-schema.

**Comment by fabian-hiller** [comment]
I will try to schedule a call with the Standard Schema team to discuss JSON Schema support again.

**Comment by kwaa** [comment]
Hello, five months after its release, I've re-summarized [xsSchema](https://xsai.js.org/docs/packages-top/xsschema):

- Weekly downloads 50k (mostly thanks to `fastmcp`)
- Zod (v3, v4, v4 mini), Valibot, ArkType, Effect Schema, and Sury (aka ReScript Schema) are supported.
- Asynchronous and synchronous interfaces
- 2.9KB minified (1.1KB gzipped)
- Maintain actively

Feel free to try xsSchema until an official solution is released, and in the meantime I'm still happy to contribute it to the Standard Schema org.

```ts
import { toJsonSchema } from 'xsschema'
```

**Comment by colinhacks** [comment]
We've added a new sister spec called **Standard JSON Schema** to address these use cases. #134
